### ISSUE TITLE
Type instability in Tape?

### ISSUE THREAD

**Issue by Sleort** [dfdx/Yota.jl#21]
I tried out the usage example
```
mutable struct Linear{T}
    W::AbstractArray{T,2}
    b::AbstractArray{T}
end

forward(m::Linear, X) = m.W * X

loss(m::Linear, X) = sum(forward(m, X))

m = Linear(rand(3,4), rand(3))
X = rand(4,5)

val, g = grad(loss, m, X)
```
and got a little surprised that a `Float32` has found its way into the tape (`const %7`):
```
julia> g.tape
Tape
  inp %1::Linear{Float64}
  inp %2::Array{Float64,2}
  const %3 = W::Symbol
  %4 = getproperty(%1, %3)::Array{Float64,2}
  %5 = *(%4, %2)::Array{Float64,2}
  %6 = sum(%5)::Float64
  const %7 = 1.0::Float32
  %8 = Yota.sum_grad(%5, %7)::Array{Float64,2}
  %9 = transpose(%2)::LinearAlgebra.Transpose{Float64,Array{Float64,2}}
  %10 = *(%8, %9)::Array{Float64,2}
  %11 = transpose(%4)::LinearAlgebra.Transpose{Float64,Array{Float64,2}}
  %12 = *(%11, %8)::Array{Float64,2}
```
Although a minor issue (which the compiler probably can take care of, at least in this example), i suspect this is a bug, right?

**Comment by dfdx** [comment]
It's not very neat, but it's intended - when mixing data of different bitness, the argument with smaller number of bits is promoted to the type with larger one. E.g. if somewhere in your code you have: 

```
1.0f0 * rand(100, 200)
```
1.0f0 will be promoted to `Float64` in almost no time. However, if you have:

```
1.0 * rand(Float32, 100, 200)
```
the array will be promoted to `Array{Float32}`, which is much, much longer operation. 

Initially I encountered this issues with `CuArrays` for which conversion to `Float64` often made the code 10x slower just because of unnecessary conversions. 

The best way to handle it, however, would be use something like `one(eltype(x))` when possible, so diffrules might get an update :)

**Comment by Sleort** [comment]
Yes, for the sake of generality/flexibility, I think you should do the latter, i.e., use `one(eltype(x))`. (I'm sure you are aware of the [performance tips](https://docs.julialang.org/en/v1/manual/performance-tips/#Write-%22type-stable%22-functions-1)...) I mean, what if someone would like to use `Float16`s? ;-)

**Comment by improbable-22** [comment]
Taking that to the logical extreme, some packages use `true` for this purpose.

**Comment by dfdx** [comment]
Fixed by #31. There are still edge cases like type-unstable zero gradient for some functions (e.g. `length()`), but they should be harmless. 

Thanks for keeping an eye on Yota's performance! Let me know if you see any other type-related issues.
